### ISSUE TITLE
Support Microsoft Guidance

### ISSUE THREAD

**Issue by vmajor** [marella/ctransformers#13]
I am trying to use a 'custom tokenizer' but I am unable to see how can I invoke it. Also can we use a standard tokenizer from HF by pulling it or loading from the local path?

**Comment by vmajor** [comment]
Never mind, I can just load it from transformers...

**Comment by marella** [comment]
Yes, custom/HF tokenizer can be used with the [`generate()`](https://github.com/marella/ctransformers#method-llmgenerate) method:

```py
from ctransformers import AutoModelForCausalLM
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('gpt2')
llm = AutoModelForCausalLM.from_pretrained('marella/gpt-2-ggml')

tokens = tokenizer.encode('AI is going to')

for token in llm.generate(tokens):
    print(tokenizer.decode(token))
```

Please let me know if you were able to use this library with your custom/HF tokenizer.

**Comment by vmajor** [comment]
I am actually trying to use ctransformers with Microsoft guidance, but I am encountering an error with protobuf. I posted a bug report there, but I am unsure what is happening.

```
import guidance
from ctransformers import AutoModelForCausalLM
from transformers import LlamaTokenizer
from transformers import AutoTokenizer

path = '/home/xxxx/models/gpt4-alpaca-lora_mlp-65B'
llm = AutoModelForCausalLM.from_pretrained('/home/xxxx/models/gpt4-alpaca-lora_mlp-65B/gpt4-alpaca-lora_mlp-65B.ggmlv3.q5_1.bin', model_type='llama')
tokenizer = AutoTokenizer.from_pretrained('decapoda-research/llama-65b-hf')
#print(llm('AI is going to'))
guidance.llm = guidance.llms.transformers.LLaMA(tokenizer, device_map='auto')
```

```
"name": "TypeError",
	"message": "Descriptors cannot not be created directly.\nIf this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.\nIf you cannot immediately regenerate your protos, some other possible workarounds are:\n 1. Downgrade the protobuf package to 3.20.x or lower.\n 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).\n\nMore information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates",
```

Downgrading protobuf to 3.2.0 just leads to another error, duplicate name.

I am not sure however if the error is with any of the components or due to the mismatch between the tokenizer and the model. The model I am using does not have a tokenizer so I am using a generic LLaMA tokenizer to feed the guidance function.

**Comment by marella** [comment]
I haven't used the guidance library but the `guidance.llms.transformers.LLaMA` class is expecting HF transformers object but you are passing ctransformers object, so it won't work.
It looks like there is already an issue opened to support ggml models https://github.com/microsoft/guidance/issues/58

**Comment by vmajor** [comment]
Yes, I saw that thread, but progress slowed down. I would really like the ability to leverage open source community efforts with what comes out of commercial, or well funded (eg HF) groups. ctransformers to me sounds like a great way to achieve that, but perhaps I am misunderstanding the drivers and philosophy behind it. Ideally I would love to be able to do what the other people in the thread were saying, just drop in a local quantized model in place of HF hosted model. This way us independent users, developers and tinkerers can plug into the much better resourced projects.

**Comment by marella** [comment]
I would also like to add support for it but it doesn't seem to have documentation on how to add new models.
I will try to follow this [example](https://github.com/microsoft/guidance/issues/58#issuecomment-1555348871) and see if I can make it work. I will look into it next weekend.

**Comment by bluecoconut** [comment]
@marella Any update on this? I'm looking forward to using StarChat-ggml weights in guidance via ctransformers~ 

I will take a stab at this later this week, but I don't want to repeat work, especially given the assumption that you might have spent time on this already. Were there any gotchas or big difficulties that I could maybe help with?

**Comment by marella** [comment]
Hey, I implemented a 🤗 Transformers compatible model and tokenizer using ctransformers and was able to run one of the examples but I think it has some bugs. I will push the code to GitHub later this week and will let you know. I'm trying to make it work like this:

```py
model = # ctransformers model
tokenizer = # ctransformers tokenizer
llm = guidance.llms.Transformers(model=model, tokenizer=tokenizer)
```

**Comment by bluecoconut** [comment]
Hey @marella, any update on this? I like the idea of having a transformers compatible model and tokenizer object for using ctransformers. I'd love to try it out, if you make a public branch with your work (even with the bugs), I can jump off from your starting point and see where it fails for me and offer some fixes if that's helpful.

**Comment by marella** [comment]
Hi, I pushed the changes to `guidance` branch. You can install using:

```sh
git clone https://github.com/marella/ctransformers
cd ctransformers
git checkout guidance
pip install -e .
```

and use it as:

```py
import guidance
from ctransformers import AutoModelForCausalLM
from ctransformers.transformers import Model, Tokenizer

llm = AutoModelForCausalLM.from_pretrained("marella/gpt-2-ggml")
tokenizer = Tokenizer(llm)
model = Model(llm)
llm = guidance.llms.Transformers(model=model, tokenizer=tokenizer)
```

I fixed some of the bugs. It is working (finishing without errors) with `guidance` `0.0.61` but getting stuck in latest version.

**Comment by qeternity** [comment]
Thanks very much for this @marella - for the life of me though I can't figure out why it hangs around 450 tokens?

**Comment by qeternity** [comment]
Ok nevermind, it appears to be an issue with `ctransformers_llm_context_length` returning an incorrect 512 for a llama model. I've overridden and everything is working now.

In doing some benchmarking `llm.eval` is much slower than llama.cpp

**Comment by kw2828** [comment]
@marella thanks for putting a branch together. I quickly tried to put together a prototype with it with your HF model above `marella/gpt-2-ggml` and `guidance==0.0.61` but it's hanging around `_stream_then_save`. Any thoughts / tweaks I can make? 

Here's a colab to reproduce: https://colab.research.google.com/drive/1YzBvp97pLwAdfl7tlKwCtYH2DZhigXiI?usp=sharing

**Comment by Jchang4** [comment]
@marella seconded! This would be killer

**Comment by lucasjinreal** [comment]
Hello, would like address some decode on Japanese character with wrong when decode in streaming.
Does anybody can help how to fix it? (this happened because of llama tokenizer vocab size small, Japanese characters need more than one token to decode correctly)

for example:
��当时年少，��

**Comment by barinov274** [comment]
> Hi, I pushed the changes to `guidance` branch. You can install using:
> 
> ```shell
> git clone https://github.com/marella/ctransformers
> cd ctransformers
> git checkout guidance
> pip install -e .
> ```
> 
> and use it as:
> 
> ```python
> import guidance
> from ctransformers import AutoModelForCausalLM
> from ctransformers.transformers import Model, Tokenizer
> 
> llm = AutoModelForCausalLM.from_pretrained("marella/gpt-2-ggml")
> tokenizer = Tokenizer(llm)
> model = Model(llm)
> llm = guidance.llms.Transformers(model=model, tokenizer=tokenizer)
> ```
> 
> I fixed some of the bugs. It is working (finishing without errors) with `guidance` `0.0.61` but getting stuck in latest version.

mistral doesn't work with guidance
Model type 'mistral' is not supported.
